### DISCUSSION TITLE
Clojure REPL: Add Support for Clojure's nREPL

### DISCUSSION THREAD

**Discussion by potapenko** [zed-industries/zed#23938] [Feature Requests]
### Summary:
Integration of Clojure REPL for Enhanced Development Experience

**Title:** Add Support for Clojure REPL to Enhance Clojure Development

**Description:**

Greetings, Zed development team,

I recently explored Zed for my Clojure projects and was truly impressed by its capabilities. The tool's intuitive design and robust feature set prompted me to consider transitioning my entire Clojure workflow into Zed. However, during my exploration, I identified a critical gap that significantly impacts the usability of Zed for Clojure development - the absence of Clojure REPL integration.

**The Issue:**

Clojure development is heavily reliant on the Read-Eval-Print Loop (REPL), a fundamental tool that enables developers to interactively write and test their code. The ability to connect to an already running REPL or initiate a new REPL session directly within the editor is not just a convenience; it's a necessity. This feature facilitates the execution of entire files (namespaces) or selective code snippets, allowing developers to immediately observe and analyze the outcomes of their code execution.

**The Request:**

I kindly urge the Zed development team to consider integrating Clojure REPL support into the editor. Such integration should enable:

- Connection to an existing REPL session or initiation of a new session within Zed.
- Execution of entire Clojure files (namespaces) or specific code snippets directly from the editor.
- Real-time observation of code execution results within the editor environment.

For a practical example of this functionality, please refer to how the [Spacemacs Clojure layer REPL](https://develop.spacemacs.org/layers/+lang/clojure/README.html#starting-a-repl-from-spacemacs) is implemented. This feature set is a cornerstone of Spacemacs' popularity among Clojure developers, despite common performance grievances.

**Why This Matters:**

Spacemacs, a popular editor among Clojure developers, already incorporates these functionalities, resulting in its widespread adoption within the Clojure community. However, Spacemacs users often report performance issues, particularly on Mac systems, which are prevalent among our community. By integrating Clojure REPL support, Zed has the potential to attract a significant portion of these users, enhancing their development experience while addressing performance concerns.

**Conclusion:**

The integration of Clojure REPL would mark a significant step forward in making Zed a comprehensive tool for Clojure development. It addresses a crucial need within the community and could dramatically increase Zed's adoption among Clojure developers. Thank you for considering this request, and I am excited about the potential improvements in future versions of Zed.

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="1128" alt="image" src="https://github.com/zed-industries/zed/assets/612926/65188f43-9c0e-47c6-9bf8-734d1bf22707">

**Comment by olaven** [comment]
👋 Chiming in to show support for this request! This is the main thing holding me back from using zed as my "general editor".

**Comment by rgbkrk** [comment]
We just added [REPL support](https://zed.dev/docs/repl) to [Zed Preview](https://zed.dev/releases/preview) through Jupyter kernels. Let me know if [IClojure](https://github.com/HCADatalab/IClojure) works for your needs or if there's another kernel you'd recommend.

**Comment by kernelp4nic** [comment]
> Let me know if [IClojure](https://github.com/HCADatalab/IClojure) works for your needs or if there's another kernel you'd recommend.

In my case it didn't work (doesn't connect to the kernel) but https://github.com/clojupyter/clojupyter worked properly

![CleanShot 2024-07-29 at 15 43 26](https://github.com/user-attachments/assets/10d47c30-9afa-48ee-a128-dbb5b5894707)
![CleanShot 2024-07-29 at 15 43 36](https://github.com/user-attachments/assets/c0f6c90c-e3f9-409e-9ad4-c27f803617b4)

I'm going to try to include clojupyter as a lib of our app to see if I can integrate the REPL inside our env.
Thanks for your work!

**Comment by rgbkrk** [comment]
Sweet. I'm going to work on getting more logs visible for end users without having to run a development copy. Hopefully we can figure out why some kernels work and some don't.

**Comment by shivangraina** [comment]
Any updates on this? Thanks.

**Comment by guigomesa** [comment]
Hey any update here? thanks

**Comment by rgbkrk** [comment]
There are more logs available in the Zed log, which you can open via the command palette. I'll need some help with someone who is (or can become) familar with IClojure.

**Comment by niquola** [comment]
I would love to see first-class support for nREPL (https://github.com/nrepl/nrepl), rather than IClojure, as that’s not how we typically work with Clojure code.

https://docs.cider.mx/cider/index.html is an idiomatic repl for clojure.

Can zed's REPL feature be extended to use nREPL?

PS:
Just an illustration how REPL driven development work in Clojure - https://www.parens-of-the-dead.com/

**Comment by tombarys** [comment]
+1 here for nREPL support. Thanks!

**Comment by franks42** [comment]
Unfortunately, without proper integrated repl (nREPL) support, ZED is not a viable coding environment for clojure... please...

**Comment by BorisVSchmid** [comment]
> > Let me know if [IClojure](https://github.com/HCADatalab/IClojure) works for your needs or if there's another kernel you'd recommend.
> 
> In my case it didn't work (doesn't connect to the kernel) but https://github.com/clojupyter/clojupyter worked properly
> 
> ![CleanShot 2024-07-29 at 15 43 26](https://private-user-images.githubusercontent.com/519481/353193883-10d47c30-9afa-48ee-a128-dbb5b5894707.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNDExOTIsIm5iZiI6MTczMzA0MDg5MiwicGF0aCI6Ii81MTk0ODEvMzUzMTkzODgzLTEwZDQ3YzMwLTlhZmEtNDhlZS1hMTI4LWRiYjViNTg5NDcwNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQwODE0NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYzhhZDcxZjhiMTBlMzhiMTJkZGMxOTEzOTlkMjk5ZWIyYzllN2FlZTYwY2M4ZDViNzE3ZjMyMDIzOTJjMWViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-ST4VYOv41SslOdwMEz_XKewR7KjOsZxvPzL2FNCMDk) ![CleanShot 2024-07-29 at 15 43 36](https://private-user-images.githubusercontent.com/519481/353193937-c0f6c90c-e3f9-409e-9ad4-c27f803617b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNDExOTIsIm5iZiI6MTczMzA0MDg5MiwicGF0aCI6Ii81MTk0ODEvMzUzMTkzOTM3LWMwZjZjOTBjLWUzZjktNDA5ZS05YWQ0LWMyN2Y4MDM2MTdiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQwODE0NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MzJlNWNkMGM0ZmViNzNhMTQ2ODBhYTg0MTUwMWEyMTA2YmRjOTBkOTM1ZWE5YjYzMzQwNjAxMmJmNGQ4NTA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4PCJl0KDcvIoeeuAw7fZBNcz5bi0agaz0N7y2YNYGg8)
> 
> I'm going to try to include clojupyter as a lib of our app to see if I can integrate the REPL inside our env. Thanks for your work!

@kernelp4nic There are multiple methods of using clojupyter - one deprecated, one on the wishlist, and two currently active. Which of these did you use?

https://github.com/clojupyter/clojupyter/blob/master/doc/usage-scenarios.md

**Comment by agriffis** [comment]
Could I recommend moving discussion of clojupyter to a separate issue? This issue is about clojure-native repl/nrepl needed for traditional clojure development, and it only muddies the water to discuss jupyter-based solutions.

**Comment by rgbkrk** [comment]
> Could I recommend moving discussion of clojupyter to a separate issue? 

I've updated the title to make it more explicit about supporting nrepl.

**Comment by guigomesa** [comment]
+1 here for nREPL support. Thanks!

**Comment by probably-neb** [comment]
Hey! This is definitely an interesting idea and something worth discussing further. We've decided to move feature/enhancements like this to discussions so I'm going to move this there. 
Feel free to continue voicing your support for adding this to Zed.

**Reply by victorb** [reply]
Here voicing my support. Been test-driving Zed a bit, but nREPL connection is a must for my day-to-day Clojure programming, so I could give it a chance to replace (neo)vim for me if nREPL was supported.

**Reply by duanebester** [reply]
Same here - I'd love nREPL support

**Reply by alexandrelasouza** [reply]
nREPL! nREPL! nREPL!

**Reply by egri-nagy** [reply]
I can do Clojure development without syntax-highlighting, auto-completion, indentation... but not without the REPL!

**Reply by MarkJacobs** [reply]
Clojure and Zed are an ideal pairing, but we do need nREPL support. Ta very much.

**Reply by agriffis** [reply]
It's not really possible for Clojure devs to even consider Zed without nREPL support.

**Reply by Dregs-2** [reply]
nREPL! nREPL! nREPL!

**Reply by shiunko** [reply]
any progress?😊

**Reply by sarnowski** [reply]
I was using Zed today to bootstrap a new Clojure service. Since Clojure is also not new anymore, the LLMs pick it up easily and it was a breeze to refactor and add features via the Agent Panel (using Claude 3.7 Sonnet fwiw).

To just echo and ping this here again: the only feature missing is a first-class integrated REPL to directly interact with your running Clojure application! @potapenko is spot on with details in the feature request. Nothing new to invent, but the execution matters.

**Reply by awb99** [reply]
Nrepl please!!

**Comment by dias-wagner** [comment]
+1 for nREPL support!

**Comment by glabrego** [comment]
+1 for nREPL support!

**Comment by functorcode** [comment]
+1 for nREPL

**Reply by JJ-Atkinson** [reply]
I want nREPL too, but constantly spamming this discussion isn't going to help ma dude.

**Comment by MikeJCusack** [comment]
There's clearly support for this being implemented. What's standing in the way of moving forward?

**Comment by kendagriff** [comment]
+1 support for this!

**Comment by pkanane** [comment]
It will be amazing to have the nREPL support 🙏

**Comment by tbrooke** [comment]
Please add nREPL support

**Comment by alanforr** [comment]
I favour adding nrepl support but the link provided in the original ticket to "Spacemacs Clojure layer REPL" is broken so I'm going to drop [a link to nrepl itself](https://github.com/nrepl/nrepl) so it's a bit easier for somebody to pick this up.

**Comment by psychedel** [comment]
My attempt to build some tools for Clojure in Zed
https://github.com/psychedel/zed-clojure-repl

It uses 
https://github.com/bhauman/rebel-readline
and
https://github.com/djblue/portal
through the Zed tasks and some scripts

I work using it in parallel with Clojure skill for Claude Code - so nREPL and project is typically aditionally controlled from the skill.

PRs are welcome

**Comment by 5icheng7hang** [comment]
plz, this is such a major pain point. i real want to write clojure in zed

**Reply by alanforr** [reply]
Have you looked at https://github.com/psychedel/zed-clojure-repl ?

**Comment by duanebester** [comment]
Taking a stab at getting this started:
https://github.com/zed-industries/zed/pull/55222

**Reply by Alounv** [reply]
Just wanted to support and share how much enthousiastic I am that you are trying to intergrate it to Zed.

Many people on this thread say they can't move to Zed without this, but my situation is opposite. I already am an everyday Zed user and I trying to learn Clojure more and more but without an integrated REPL it feels like a good chunk of what makes this language exiting is lost.

So 💪🏼 👏🏼

**Comment by MikeJCusack** [comment]
Any movement on this? There seems to be a lot of interest in getting this working.

**Reply by alanforr** [reply]
Have you looked at https://github.com/psychedel/zed-clojure-repl ?

**Reply by aggsol** [reply]
This is a draft of an idea.

**Comment by aggsol** [comment]
Would setting up a bounty help?

**Comment by duanebester** [comment]
@josephinoo has taken my draft work & PR to new heights - https://github.com/zed-industries/zed/pull/63219
